src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -520,6 +520,14 @@ impl Step for Miri {
         builder.ensure(compile::Std::new(target_compiler, host));
         let host_sysroot = builder.sysroot(target_compiler);
 
+        // Miri has its own "target dir" for ui test dependencies. Make sure it gets cleared
+        // properly when rustc changes. Similar to `Builder::cargo`, we skip this in dry runs to
+        // make sure the relevant compiler has been set up properly.
+        if !builder.config.dry_run() {
+            let ui_test_dep_dir = builder.stage_out(host_compiler, Mode::ToolStd).join("ui");
+            builder.clear_if_dirty(&ui_test_dep_dir, &builder.rustc(host_compiler));
```

Comment:
What is that "target dir" called? It feels like this is going to affect other tests; it doesn't seem miri-specific logic.

Reply:
This ends up being `build/host/stage0-tools/ui` on my system. It is true that this is not Miri-specific. There is a `miri` subfolder inside this that we could clear, but strangely there is also `build/host/stage0-tools/ui/tests` with a whole bunch of empty directories (one for each directory in the Miri test suite), so Miri's artifacts are not contained to the `miri` subfolder.

AFAIK ui_test picks its output dir inside the "target dir" of the project it is being built for, so when a build environment puts many projects into the same "target dir" things can start becoming confusing... the logic for that is [here](https://github.com/oli-obk/ui_test/blob/b108b5fe1d35acb4c818b521b55bd71416aa42fb/src/config.rs#L144-L147).
Cc @oli-obk